docs/proposals/TLSPolicy.md

```diff
@@ -0,0 +1,84 @@
+
+**Choosing a TLS Strategy**: As an admin, I want to employ different strategies for managing certificates (e.g., multiple host certificate or per host certificate).
+
+**Choosing a TLS Strategy**: As an admin, I want to be able to apply enterprise security standards to certificates generated by the MGC. I want to be able to specify certificate key size, algorithm and encoding.
+
+
+## Current implementation
+
+
+
+## Goals
+
+- Design an API that allows admins to handle Automatic TLS for any given gateway.
+- Facilitate easy selection of a TLS provider.
+- Allow admins to set the renewal window for TLS certificates.
+- Enable the use of different TLS strategies.
+- Consider additional TLS configurations such as enforcing HTTPS listeners and supporting wildcard hosts.
+
+## Non-Goals
+
+- Detailed implementation of the API.
+- Developing a working prototype of the API.
+- Identifying specific technologies or tools for implementing the API.
+- Abstracting cert-manager Issuer/ClusterIssuer configuration.
+
+## Terms
+
+- **Automatic TLS**: Automatic management of TLS certificates for gateways.
+- **TLS Provider**: The entity providing TLS certificates.
+- **TLS Strategy**: The method used for managing and distributing TLS certificates across hosts.
+
+## Proposal
+
+The proposed solution is the TLSPolicy API. It allows administrators to select a TLS provider, define a renewal window for TLS certificates, and choose a TLS strategy. The API will support multiple host certificates and per-host certificates.
+
+The API will also consider additional TLS configurations, such as enforcing only HTTPS listeners and supporting wildcard hosts. This means any gateways using this class would have any non-HTTPS listeners flagged, removed, or not synced, etc. 
+
+**Note** This proposal effectively moves us away from having a "default" TLS policy, baked into the MGC. From here on, unless a TLSPolicy is defined, MGC will not do anything with TLS, and control will remain with the user to solve.
+
+## Gateway Interaction
+
+The TLSPolicy API will interact with Gateway operations, such as adding and removing TLS configurations for a listener based on the host. This allows for dynamic management of TLS configurations for each Gateway and its listeners. The API will utilize the existing Gateway methods for adding, removing, and checking TLS status.
+
+## Example Policies
+
+1. **Policy**
+
+```yaml
+apiVersion: kuadrant.io/v1alpha1
+kind: TLSPolicy
+spec:
+  targetRef: # defaults to gateway gvk and currrent namespace
+    name: gateway-name
+  duration: 2160h # 90d - certificate validity period
+  renewBefore: 360h # 15d - certificate renewal buffer in hours
+  PrivateKey:
+    Algorithm: RSA # certman.RSAKeyAlgorithm
+    Encoding: PKCS1 # certman.PKCS1
+    Size: 2048
```

Comment:
hmm expected to see an issuer property here?

Reply:
My understanding would have led me to expect the same. I almost feel like that's the main thing we want to allow people to select through the policy.

TLSPolicy in dev just references the self signed cert issuer, or are we expecting that the policy would manage the creation of everything for you? :thinking: This might be nice, but probably completely unrealistic for a first pass, the options we would have to provide to create a vaid Issuer in cert manager would be huge .

Reply:
yes self signed in dev. Issuer is expected with this policy as it stands

Reply:
I'll drop an updated copy of this when I generate the CRDs